sale.py

```diff
@@ -0,0 +1,226 @@
+
+    @classmethod
+    def __setup__(cls):
+        """
+        Setup the class before adding to pool
+        """
+        super(Sale, cls).__setup__()
+
+        cls._sql_constraints += [
+            (
+                'amazon_id_unique',
+                'UNIQUE(amazon_order_id)',
+                'Sale with this amazon Order ID already exists'
+            )
+        ]
+
+    @classmethod
+    def find_or_create_using_amazon_id(cls, order_id):
+        """
+        This method tries to find the sale with the order ID
+        first and if not found it will fetch the info from amazon and
+        create a new sale with the data from amazon using
+        create_using_amazon_data
+
+        :param order_id: Order ID from amazon
+        :type order_id: string
+        :returns: Active record of sale order created/found
+        """
+        SaleChannel = Pool().get('sale.channel')
+
+        sales = cls.search([
+            ('amazon_order_id', '=', order_id),
+        ])
+        if sales:
+            return sales[0]
+
+        amazon_channel = SaleChannel(
+            Transaction().context.get('amazon_channel')
+        )
+        order_api = amazon_channel.get_order_api()
+
+        order_data = order_api.get_order([order_id]).parsed
+
+        order_line_data = order_api.list_order_items(
+            order_data['Orders']['Order']['AmazonOrderId']['value']
+        ).parsed
+
+        return cls.create_using_amazon_data(
+            order_data['Orders']['Order'],
+            order_line_data['OrderItems']['OrderItem']
+        )
+
+    @classmethod
+    def create_using_amazon_data(cls, order_data, line_data):
+        """
+        Create a sale using amazon data
+
+        :param order_data: Order data from amazon
+        :return: Active record of record created
+        """
+        Party = Pool().get('party.party')
+        Address = Pool().get('party.address')
+        SaleChannel = Pool().get('sale.channel')
+        Currency = Pool().get('currency.currency')
+        Uom = Pool().get('product.uom')
+
+        amazon_channel = SaleChannel(
+            Transaction().context.get('amazon_channel')
+        )
+
+        currency, = Currency.search([
+            ('code', '=', order_data['OrderTotal']['CurrencyCode']['value'])
+        ], limit=1)
+
+        # Order lines are returned as dictionary for single record and as list
+        # for mulitple reocrds.
+        # So convert to list if its dictionary
+        if isinstance(line_data, dict):
+            # If its a single line order, then the array will be dict
+            order_items = [line_data]
+        else:
+            # In case of multi line orders, the transaction array will be
+            # a list of dictionaries
+            order_items = line_data
+
+        party_values = {
+            'name': order_data['BuyerEmail']['value'],
+            'email': order_data['BuyerName']['value'],
+        }
+        party = Party.create_using_amazon_data(party_values)
```

Comment:
this should be a find_or_create call especially when you have amazon_email as a unique identifier for the party.